### DISCUSSION TITLE
How to fetch association from field datafetcher

### DISCUSSION THREAD

**Discussion by marceloverdijk** [Netflix/dgs-framework#21] [Q&A] [answered]
This is a bit of a follow up to https://github.com/Netflix/dgs-framework/discussions/18 but the question is different.

Let's take this example:

```
type Query {
    shows: [Show]
}

type Show {
   title: String
   director: Director
}

type Director {
   name: String
}
```

The codegen generates a `Director` class and a `Show` class and the latter has a `getDirector()` method.

And the implementing data fetchers could be something like:

```
@DgsData(parentType = "Query", field = "shows"
List<Show> shows() {
    // Get list of shows from persistent layer, but without the director.
    // We also convert the show from the persistent model class to the generated DGS type.
    // Note that the persistent Show model class has a reference to the Director (like a `directorId`, which cannot be stored on the generated type)
    var shows = ...;
    return shows;
}

@DgsData(parentType = "Show", field = "director")
Director director(DataFetchingEnvironment dfe) {
    // How to get the director now?
    // dfe.getSource() would return the codegen generated `Show` which has no reference to the `directorId`.
    var director = ...
}
```

In the director data fetcher I could retrieve the persistent Show (again) - using like `showRepository.getShow(dfe.getSource().getId())` - and from there get the `directorId` and query the Director.

I wonder if there is a better way?

PS: Without DGS my Show class returned by the shows datafetcher would contain the `directorId` but would not be exposed in the GraphQL schema. With the codegen generated classes this is not possible, so I'm looking for a nice pattern to implement this.

**Comment by paulbakker** [comment, marked as answer]
I think I understand the problem; to be able to load the director you need a field that your internal datas structure has, but is not represented in the schema, and thus not on the generated type.
If you return an instance of the generated type from the `shows` data fetcher, that's what you get back from `getSource`.

This problem is introduced because the `director` data fetcher requires context that's only available in the `shows` data fetcher. I'm sure there are multiple ways to get this to work, but my first approach would be to create a [DataLoader](https://netflix.github.io/dgs/data-loaders/) to load directors, and invoke that data loader from within the `shows` data fetcher where you still have the `directorId`.

**Reply by marceloverdijk** [reply]
> I think I understand the problem; to be able to load the director you need a field that your internal datas structure has, but is not represented in the schema, and thus not on the generated type.

Indeed.

> I'm sure there are multiple ways to get this to work, but my first approach would be to create a DataLoader to load directors, and invoke that data loader from within the `shows` data fetcher where you still have the `directorId`.

With this approach I would need to check in the ` shows` data fetcher if the `director` field was requested as I don't want to fetch this if not requested (in the example I used a Director, but in practice it could also be an expensive time consuming call). But it would also become a hassle if there are a dozen of these kind of associations on the `Show` type, or think about a more recursive schema where the `Director` type itself has these kind of associations like a `countryOfBirth`, `nationality` etc.

Maybe I should use the GraphQL local context to push down the data down to the child data fetchers instead.
Like discussed here: https://spectrum.chat/graphql-java/general/how-does-local-context-work~6f3adccf-8d1b-47c2-9233-23310a183903
E.g. storing the persistent model instance in the local context, so the `director` data fetcher can retrieve it and access the `directorId` field from it.

Does DGS support returning a `DataFetcherResult` instance from the data fetcher instead of a generated type?

**Reply by srinivasankavitha** [reply]
Yes, you can return a DataFetcherResult from a data fetcher. For additional context: https://www.graphql-java.com/blog/deep-dive-data-fetcher-results/

**Reply by marceloverdijk** [reply]
Thx @srinivasankavitha for the link.

I never noticed that page in the `graphql-java` docs as it is under the blog section.
And yes as I thought the local context indeed seems a way to pass data down the child data fetchers.

However, when using it in combination with data loaders I'm still having a problem.
But to explain it I need to provide more details.

Here is a part of my persistent model.

```
public class PersonModel {
    String id;
    String name;
    String countryOfBirthCountryId;
    String nationalityCountryId;
}

public class CountryModel {
    String id;
    String code;
    String name;
    String continentId;
}

public class ContinentModel {
    String id;
    String name;
}
```

and the GraphQL schema looks like:

```
Query {
    person(id: ID!): Person
}

type Person {
    id: ID!
    name: String!
    countryOfBirth: Country
    nationality: Country!
}

type Country {
    id: ID!
    code: String!
    name: String!
    continent: Continent!
}

type Continent {
    id: ID!
    name: String!
}
```

With that in mind I created the data fetchers as follow:

```
@DgsComponent
@RequiredArgsConstructor
public class PersonDatafetcher {

    private final PersonModelRepository personModelRepository;
    private final PersonModelMapper personModelMapper;

    @DgsData(parentType = DgsConstants.QUERY_TYPE, field = DgsConstants.QUERY.Person)
    public DataFetcherResult<Person> person(@InputArgument("id") String id) {
        var personModel = personModelRepository.findById(id);
        return DataFetcherResult.<Person>newResult()
                .data(personModelMapper.toPerson(personModel))
                .localContext(personModel) // Store the person model to retrieve countryOfBirthCountryId and nationalityCountryId later in child data fetchers.
                .build();
    }

    @DgsData(parentType = DgsConstants.PERSON.TYPE_NAME, field = DgsConstants.PERSON.CountryOfBirth)
    public CompletableFuture<Country> countryOfBirth(DgsDataFetchingEnvironment dfe) {
        PersonModel personModel = dfe.getLocalContext();
        if (personModel.getCountryOfBirthCountryId() == null) {
            return null;
        }
        DataLoader<String, Country> dataLoader = dfe.getDataLoader(CountriesDataLoader.class);
        return dataLoader.load(personModel.getCountryOfBirthCountryId());
    }

    @DgsData(parentType = DgsConstants.PERSON.TYPE_NAME, field = DgsConstants.PERSON.Nationality)
    public CompletableFuture<Country> nationality(DgsDataFetchingEnvironment dfe) {
        PersonModel personModel = dfe.getLocalContext();
        if (personModel.getNationalityCountryId() == null) {
            return null;
        }
        DataLoader<String, Country> dataLoader = dfe.getDataLoader(CountriesDataLoader.class);
        return dataLoader.load(personModel.getNationalityCountryId());
    }
}
```

and the countries data loader:

```
@DgsDataLoader(name = "countries")
@RequiredArgsConstructor
public class CountriesDataLoader implements BatchLoader<String, Country> {

    private final CountryModelRepository countryModelRepository;
    private final CountryModelMapper countryModelMapper;

    @Override
    public CompletionStage<List<Country>> load(List<String> keys) {
        return CompletableFuture.supplyAsync(() -> keys
                .stream()
                .map(id -> countryModelRepository.findById(id))
                .map(countryModelMapper::toCountry)
                .collect(Collectors.toList()));
    }
}
```

So far this works.
With the `PersonModel` set in the local context the `countryOfBirth` and `nationality` data fetcher can retrieve the appropriate country id's and send them to the countries data loader.

Now we can query like:

```
{
  person(id: "jason-bateman") {
    name
    countryOfBirth {
      code
      name
    }
    nationality {
      name
    }
  }
}
```

but I'm having a problem querying the continent:

```
{
  person(id: "jason-bateman") {
    name
    countryOfBirth {
      code
      name
      continent {
        name
      }
    }
    nationality {
      name
    }
  }
}
```

The countries are fetched using a data loader - which as far as I can see - cannot change the local context.

So creating a continent data fetcher like:

```
@DgsComponent
@RequiredArgsConstructor
public class ContinentDatafetcher {

    private final ContinentModelRepository continentModelRepository;
    private final ContinentModelMapper continentModelMapper;

    @DgsData(parentType = DgsConstants.COUNTRY.TYPE_NAME, field = DgsConstants.COUNTRY.Continent)
    public DataFetcherResult<Continent> continent(DgsDataFetchingEnvironment dfe) {
        String continentId = ... // How to get the continent here?
        DataLoader<String, Continent> dataLoader = dfe.getDataLoader(ContinentsDataLoader.class);
        return dataLoader.load(continentId);
    }
}
```

Sorry for the long post, but I hope to give insight what I'm trying to do, and hopefully there is a way I can solve this.
Note that the persistent model I have is an existing model I cannot change.

**Reply by paulbakker** [reply]
What about extending/composing the generated Country type with a countryId field? Fields that are set in the Java instances, but were not requested by the query are just dropped by the framework when creating the response, so you can have fields that you use internally without exposing them to your users.

**Reply by paulbakker** [reply]
To demonstrate what I said above I implemented an example based on the DGS Java example: https://github.com/Netflix/dgs-examples-java/compare/internal-type-demo. In the example I removed the `id` field from `Show` in the schema. I still need the id to load `reviews` however, so I introduced an `InternalShow` type that delegates to Show and adds an id. In the datafetcher/dataloader code it's using InternalShow, while the user will never know.  I hope that helps.

**Reply by marceloverdijk** [reply]
Yes this is a better pattern with a "backing object" containing other needed data.

This is indeed similar as I do in a vanilla `graphql-java` project.

Don't know why I did not came up myself extending/wrapping the dgs code generated classes.

Personally I prefer extending instead of wrapping but that's a personal choice, but while thinking about I would maybe even prefer more implementing an interface!

Did you ever think about having the option to generate Java interfaces complementary to the generated classes? E.g. with a `generateInterfaces` option similar as `generateClient`.
The the data fetchers would then expose these interfaces as the contract. Internally I could use the codegen generated classes (as they should implement the interfaces as well) or in situations where I need more I simply create a custom implementation; e.g. a simple `data` class.
The benefit using codegen is still huge as I'm guaranteed with the schema contract.

Wdyt? Worth to create a feature request?

**Reply by paulbakker** [reply]
Using interfaces would work. I'm personally not a huge fan of it, it feels like ending up with more types than really needed, but I'm not opposed to it either.

**Reply by marceloverdijk** [reply]
@paulbakker I tested a bit further with the DGS framework and I think my need for having interfaces being generated increased.
Not absolutely necessary, but let me explain. 

Building on my previous example, and adding `persons(page: Int = 0, size: Int = 10): PersonPage!` to the query.

```
Query {
    person(id: ID!): Person
    persons(page: Int = 0, size: Int = 10): PersonPage!
}

type Person {
    id: ID!
    name: String!
    countryOfBirth: Country
    nationality: Country!
}

type PersonPage {
    items: [Person!]!
    pageInfo: PageInfo!
}

type Country {
    id: ID!
    code: String!
    name: String!
    continent: Continent!
}

type PageInfo {
    hasNextPage: Boolean!
    hasPreviousPage: Boolean!
    totalCount: Int!
    totalPages: Int!
}
```

Like I said internally I can use a custom DTO extending the generated `Person` like:

```
@Data
@EqualsAndHashCode(callSuper = true)
public class PersonDTO extends Person {

    private String countryOfBirthCountryId;
    private String nationalityCountryId;
}
```

For the `person(id: ID!): Person` query this works fine, but for the generated `PersonPage` this won't work that nicely.
It generated:

```
public class PersonPage {

  private List<Person> items;
  private PageInfo pageInfo;

  ..
}
```

Here the list also needs to be `List<PersonDTO> items` for the same reasons to retrieve the country of birth and nationality later.
I cannot simply extend the `PersonPage` here but should create a custom class without any relation the `PesonPage` anymore, like:

```
public class PersonPageDTO { // cannot extend PersonPage to have List<PersonDTO> items

  private List<PersonDTO> items;
  private PageInfo pageInfo;

  ..
}
```

With this setup I will agree it will work, but losing the contact with the generated `PersonPage` is not what I like.


Now, if for `Person` and `PersonPage` an interface would have been generated (let's say `IPerson` and `IPersonPage` for the example) I could have used:

```
public class PersonPageDTO implements IPersonPage {

  private List<IPerson> items;
  private IPageInfo pageInfo;

  ..
}
```

and I would fully comply to the generated GraphQL schema contract of `IPersonPage`.

**Reply by driptaroop** [reply]
@paulbakker That is a nice pattern indeed. Although it would work for most of the usecases, there are some cases (like @marceloverdijk added), I would also think that generating an interface might be required. 

If we're using Kotlin, using this pattern is a bit challenging because the generated data models would be data classes which cannot be extended (since data classes are final by design). so in order to use this pattern we have cannot simply extend the generated model and add other parameters. we have to write separate data classes manually which would defeat the purpose of using the generated models as contract. In this case also it would be great to have an interface that can be extended.

**Reply by willyxiao** [reply]
Does `generateInterfaces` work for kotlin? Right now, I see that it works for `language = "java"` but I don't see any interfaces generated with `language = "kotlin"`

**Comment by paulbakker** [comment]
@marceloverdijk I've created a PR with new docs showing the different options around this topic if you want to review: https://github.com/Netflix/dgs/pull/10.

**Reply by marceloverdijk** [reply]
Yes that looks very good! I left a comment at the PR with a remark.

**Comment by torbenkeller** [comment]
To reduce the amount of boilerplate code, but still using the `InternalShow` aproach, could be schema tagging. I am thinking about something like this:

```graphql
type Show {
  @internal id: ID
  title: String
  reviews: [Review]
}
```

The framework could then ignore all fields annotated with `@internal` and the codegen plugin could add those fields to the generated data classes.
